src/3d/qgs3dmapsettings.cpp

```diff
@@ -340,6 +348,13 @@ QDomElement Qgs3DMapSettings::writeXml( QDomDocument &doc, const QgsReadWriteCon
   elemOrigin.setAttribute( QStringLiteral( "z" ), QString::number( mOrigin.z() ) );
   elem.appendChild( elemOrigin );
 
+  QDomElement elemExtent = doc.createElement( QStringLiteral( "extent" ) );
+  elemExtent.setAttribute( QStringLiteral( "xMin" ), QString::number( mExtent.xMinimum() ) );
```

Comment:
```suggestion
  elemExtent.setAttribute( QStringLiteral( "xMin" ), qgsDoubleToString( mExtent.xMinimum() ) );
```
(elsewhere too)

Reply:
Actually, conversion to string seems redundant here. I removed the string conversion and directly used the `.setAttribute( QString, double )` variant.
This should probably be also done in the other places that use the `QString::number()`, like when storing the `mOrigin` (will handle in a separate pr)

Reply:
Actually it's still needed -- it's done to avoid unwanted loss of precision when QString::number or the QDomElement internals convert large values to scientific notation.

Reply:
`QDomElement.setAttribute( QString, double )` seems to use `QString.setNum( value, 'g', 17 )` under the hood, which means _either decimal or scientific notation, whichever is more concise_.
Our `qgsDoubleToString()` uses `QString::Number( value, 'f', 17 )` to force not using scientific notation. The precision gain though is probably misleading. Here's a test to compare _'f' vs 'g'_ with large doubles:
```cpp
  QString x;
  double v = 1234567890.12345678901234567890;
  QgsDebugMsg( x.setNum( v, 'g', 17 ) );  // 1234567890.1234567
  QgsDebugMsg( x.setNum( v, 'f', 17 ) );  // 1234567890.12345671653747559
  QgsDebugMsg( qgsDoubleToString( v ) );  // 1234567890.12345671653747559
//                                                             ^--------- not useful data
  v = 12345678901234567890.12345678901234567890;
  QgsDebugMsg( x.setNum( v, 'g', 17 ) );  // 1.2345678901234567e+19
  QgsDebugMsg( x.setNum( v, 'f', 17 ) );  // 12345678901234567168.00000000000000000
  QgsDebugMsg( qgsDoubleToString( v ) );  // 12345678901234567168
//                                                            ^-- precision loss
```

I might be missing something, but to me it seems the only _real_ gain is the trimming of the trailing zeros

Reply:
Alright... digging down the rabbit hole! I think that we started using qgsDoubleToString originally instead of QDomElement.setAttribute( ..., double ) was because in qt 4 .setAttribute used the current locale  (see https://doc.qt.io/archives/qt-4.8/qdomelement.html#setAttribute-7 -- "The number is formatted according to the current locale."), vs QString::number which didn't.  But QString::number would use insufficient precision by default, so we moved wholesale to writing doubles to xml via qgsDoubleToString.

Somewhere along the way QDomElement.setAttribute behaviour has changed -- see https://doc.qt.io/qt-5/qdomelement.html#setAttribute-6 "The formatting always uses QLocale::C.".

So ultimately we can (**probably** :laughing: )  safely use setAttribute directly now instead of qgsDoubleToString!

Reply:
Wow, thanks for the analysis! I hadn't noticed that in qt4 _setAttribute_ was locale aware. That would have been terrible for storing settings :)
I'm probably leaving it as it is then ;)